File: docs/spec/hooks/examples/python/hooks_interchange/exit_codes.py
"""Exit code and decision resolution for hook handlers.

Implements the evaluation order and truth table from the
Hook Interchange Format Specification, Sections 4 and 5.

Conformance level: Extended (Section 8.2).
"""

from __future__ import annotations

from enum import Enum


class Decision(Enum):
    """Structured JSON ``decision`` values from §5.1."""

    ALLOW = "allow"
    DENY = "deny"
    ASK = "ask"


class Result(Enum):
    """Final resolved outcome after applying §5.3 evaluation order."""

    ALLOW = "allow"
    BLOCK = "block"
    WARN_ALLOW = "warn_allow"
    ASK = "ask"


def resolve(blocking: bool, exit_code: int, decision: Decision | None) -> Result:
    """Resolve the combined exit code and JSON decision into a final result.

    Applies the two-step evaluation order mandated by §5.3:

    1. **Non-blocking downgrade** (§4, §5.3 step 1): when ``blocking`` is
       ``False``, exit code 2 MUST be treated as exit code 1 before any
       further evaluation.
    2. **Truth table lookup** (§5.3 step 2): apply the precedence rules from
       §5.2 using the (possibly downgraded) exit code and ``decision``.

    Exit codes outside {0, 1, 2} are normalised to 1 per §4 ("Other: same
    behaviour as exit code 1").

    When ``decision`` is ``None`` (absent from the JSON output) and exit code
    is 0, the result is ``ALLOW`` per §5.2 ("When the ``decision`` field is
    absent and exit code is 0, the implementation MUST treat the result as
    ``decision: 'allow'``").

    Note on invalid JSON stdout: when a hook exits with code 0 and stdout is
    not valid JSON, callers MUST substitute exit code 1 before calling this
    function and pass ``decision=None``. This function does not receive raw
    stdout; JSON parsing and the exit-code substitution are the caller's
    responsibility (§5.2: "implementations MUST treat the result as exit
    code 1").

    Truth table (§5.3):

    +-----------+-----------+------------+-------------------------------+
    | blocking  | exit_code | decision   | result                        |
    +===========+===========+============+===============================+
    | True      | 0         | allow      | ALLOW                         |
    +-----------+-----------+------------+-------------------------------+
    | True      | 0         | deny       | BLOCK                         |
    +-----------+-----------+------------+-------------------------------+
    | True      | 0         | ask        | ASK                           |
    +-----------+-----------+------------+-------------------------------+
    | True      | 0         | None       | ALLOW                         |
    +-----------+-----------+------------+-------------------------------+
    | True      | 2         | allow      | BLOCK  (exit code 2 overrides)|
    +-----------+-----------+------------+-------------------------------+
    | True      | 2         | deny       | BLOCK                         |
    +-----------+-----------+------------+-------------------------------+
    | True      | 1         | any        | WARN_ALLOW                    |
    +-----------+-----------+------------+-------------------------------+
    | False     | 0         | allow      | ALLOW                         |
    +-----------+-----------+------------+-------------------------------+
    | False     | 0         | deny       | BLOCK                         |
    +-----------+-----------+------------+-------------------------------+
    | False     | 0         | None       | ALLOW                         |
    +-----------+-----------+------------+-------------------------------+
    | False     | 2         | any        | WARN_ALLOW  (downgraded to 1) |
    +-----------+-----------+------------+-------------------------------+
    | False     | 1         | any        | WARN_ALLOW                    |
    +-----------+-----------+------------+-------------------------------+

    Args:
        blocking: Whether the hook definition has ``blocking: true``.
        exit_code: The raw exit code returned by the hook process.
        decision: The parsed ``decision`` field from the JSON output, or
            ``None`` when the field is absent or stdout was empty/invalid.

    Returns:
        The resolved ``Result``.
    """
    # Normalise exit codes outside {0, 1, 2} to 1 (§4 "Other: hook error").
    if exit_code not in (0, 1, 2):
        exit_code = 1

    # Step 1 — non-blocking downgrade (§5.3 step 1).
    # When blocking is False, exit code 2 MUST be treated as exit code 1
    # before any further evaluation.
    if not blocking and exit_code == 2:
        exit_code = 1

    # Step 2 — truth table lookup (§5.3 step 2).

    # exit_code 1 (including downgraded 2) → WARN_ALLOW regardless of decision.
    if exit_code == 1:
        return Result.WARN_ALLOW

    # exit_code 2 (only reachable when blocking is True, after step 1).
    # Exit code 2 overrides any JSON decision — the action is always blocked.
    if exit_code == 2:
        return Result.BLOCK

    # exit_code 0: decision field drives the result.
    assert exit_code == 0  # noqa: S101 — unreachable otherwise

    if decision is None or decision is Decision.ALLOW:
        return Result.ALLOW
    if decision is Decision.DENY:
        return Result.BLOCK
    if decision is Decision.ASK:
        return Result.ASK

    # Exhaustive enum coverage — unreachable with a well-formed Decision value.
    raise ValueError(f"Unhandled decision value: {decision!r}")  # pragma: no cover

File: docs/spec/hooks/examples/typescript/src/exitCodes.ts
/**
 * exitCodes.ts — Exit Code Resolution
 * Reference implementation: Extended conformance (§5, §8.2)
 *
 * Covers:
 *   §4    Exit Code Contract — exit code semantics and normalization
 *   §5.2  Interaction with Exit Codes — precedence rules
 *   §5.3  Evaluation Order — non-blocking downgrade then truth table lookup
 */

// ---------------------------------------------------------------------------
// Type definitions
// ---------------------------------------------------------------------------

/** The JSON `decision` field from §5.1. `null` represents an absent field. */
export type Decision = "allow" | "deny" | "ask";

/** The resolved result after applying the full evaluation order from §5.3. */
export type Result = "allow" | "block" | "warn_allow" | "ask";

// ---------------------------------------------------------------------------
// Resolution
// ---------------------------------------------------------------------------

/**
 * Resolve the combined exit code and JSON decision into a final result.
 *
 * Implements the evaluation order defined in §5.3:
 *
 * 1. **Normalize exit code** — Any code other than 0, 1, or 2 is treated as 1
 *    (§4: "Other" exit codes have the same behavior as exit code 1).
 *
 * 2. **Non-blocking downgrade** — If `blocking` is `false` and the normalized
 *    exit code is 2, it is downgraded to 1 before further evaluation (§5.3 step 1).
 *
 * 3. **Truth table lookup** — The normalized (and possibly downgraded) exit code
 *    is combined with `decision` to produce the final result (§5.3 step 2).
 *
 * @param blocking  Whether the hook definition has `blocking: true`.
 * @param exitCode  The raw integer exit code from the hook process.
 * @param decision  The parsed `decision` field from stdout JSON, or `null` when
 *                  absent (empty stdout, `{}`, or no `decision` key).
 */
export function resolve(
  blocking: boolean,
  exitCode: number,
  decision: Decision | null,
): Result {
  // Step 1 — Normalize: map any exit code outside {0, 1, 2} to 1.
  const normalizedCode = exitCode === 0 || exitCode === 2 ? exitCode : 1;

  // Step 2 — Non-blocking downgrade: exit code 2 becomes 1 when blocking is false.
  const effectiveCode =
    !blocking && normalizedCode === 2 ? 1 : normalizedCode;

  // Step 3 — Truth table lookup (§5.3).
  if (effectiveCode === 1) {
    // exit code 1 always yields warn_allow, regardless of blocking or decision.
    return "warn_allow";
  }

  // effectiveCode is 0 from here on. Blocking only matters for exit code 2,
  // which is handled above; for code 0 the decision field drives the result.
  if (effectiveCode === 2) {
    // Only reachable when blocking === true (downgrade already applied above).
    // exit code 2 overrides decision: "allow" per §5.2.
    return "block";
  }

  // effectiveCode === 0: decision field is authoritative.
  switch (decision) {
    case "deny":
      // decision: "deny" with exit code 0 is treated as a block (§5.2).
      return "block";
    case "ask":
      // decision: "ask" defers to the user (§5.2). The caller is responsible
      // for degrading to "block" in non-interactive environments.
      return "ask";
    case "allow":
    case null:
      // decision: "allow" or absent — normal success path (§5.2).
      return "allow";
  }
}
